asserts/extkeypairmgr.go

```diff
@@ -0,0 +1,260 @@
+)
+
+// ExternalKeypairManager is key pair manager implemented via an external program interface.
+// TODO: points to interface docs
+type ExternalKeypairManager struct {
+	keyMgrPath string
+	nameToID   map[string]string
+	cache      map[string]*cachedExtKey
+}
+
+// NewExternalKeypairManager creates a new ExternalKeypairManager using the program at keyMgrPath.
+func NewExternalKeypairManager(keyMgrPath string) (*ExternalKeypairManager, error) {
+	em := &ExternalKeypairManager{
+		keyMgrPath: keyMgrPath,
+		nameToID:   make(map[string]string),
+		cache:      make(map[string]*cachedExtKey),
+	}
+	if err := em.checkFeatures(); err != nil {
+		return nil, err
+	}
+	return em, nil
+}
+
+func (em *ExternalKeypairManager) keyMgr(op string, args []string, in []byte, out interface{}) error {
+	args = append([]string{op}, args...)
+	cmd := exec.Command(em.keyMgrPath, args...)
+	var outBuf bytes.Buffer
+	var errBuf bytes.Buffer
+
+	if len(in) != 0 {
+		cmd.Stdin = bytes.NewBuffer(in)
+	}
+	cmd.Stdout = &outBuf
+	cmd.Stderr = &errBuf
+
+	if err := cmd.Run(); err != nil {
+		return fmt.Errorf("external keypair manager %q %v failed: %v (%q)", em.keyMgrPath, args, err, errBuf.Bytes())
+
+	}
+	switch o := out.(type) {
+	case *[]byte:
+		*o = outBuf.Bytes()
+	default:
+		err := json.Unmarshal(outBuf.Bytes(), out)
+		if err != nil {
+			return fmt.Errorf("cannot decode external keypair manager %q %v output: %v", em.keyMgrPath, args, err)
+		}
+	}
+	return nil
+}
+
+func (em *ExternalKeypairManager) checkFeatures() error {
+	var feats struct {
+		Signing    []string `json:"signing"`
+		PublicKeys []string `json:"public-keys"`
+	}
+	if err := em.keyMgr("features", nil, nil, &feats); err != nil {
+		return err
+	}
+	if !strutil.ListContains(feats.Signing, "RSA-PKCS") {
+		return fmt.Errorf("external keypair manager %q unexpectedly does not support RSA-PKCS", em.keyMgrPath)
+	}
+	if !strutil.ListContains(feats.PublicKeys, "DER") {
+		return fmt.Errorf("external keypair manager %q unexpectedly does not support public key DER output format", em.keyMgrPath)
+	}
+	return nil
+}
+
+func (em *ExternalKeypairManager) keyNames() ([]string, error) {
+	var knames struct {
+		Names []string `json:"key-names"`
+	}
+	if err := em.keyMgr("key-names", nil, nil, &knames); err != nil {
+		return nil, fmt.Errorf("cannot get all external keypair manager key names: %v", err)
+	}
+	return knames.Names, nil
+}
+
+func (em *ExternalKeypairManager) findByName(name string) (PublicKey, *rsa.PublicKey, error) {
+	var k []byte
+	err := em.keyMgr("get-public-key", []string{"-f", "DER", "-k", name}, nil, &k)
+	if err != nil {
+		return nil, nil, fmt.Errorf("cannot find external key: %v", err)
+	}
+	pubk, err := x509.ParsePKIXPublicKey(k)
+	if err != nil {
+		return nil, nil, fmt.Errorf("cannot decode external key %q: %v", name, err)
+	}
+	rsaPub, ok := pubk.(*rsa.PublicKey)
+	if !ok {
+		return nil, nil, fmt.Errorf("expected RSA public key, got instead: %T", pubk)
```

Comment:
Maybe "expected RSA public key, got %T instead" or "expected RSA public key, got: %T"? CC @degville

Reply:
I think Pawel is right - or maybe `received %T instead of expected RSA public key.`

Reply:
I followed the pattern of some preexisting errors in crypto.go, if we change this one I have to change the others too I think